src/commissioner/ExampleCommissioningStateMachine.h

```diff
@@ -0,0 +1,664 @@
+    void Enter()
+    {
+        uint8_t attestationNonce[chip::kAttestationNonceLength];
+        chip::Crypto::DRBG_get_bytes(attestationNonce, sizeof(attestationNonce));
+        this->mCtx.Dispatch(Event::Create<AttestationInformation>(chip::ByteSpan(attestationNonce)));
+    }
+};
+
+struct AttestationVerification : chip::Commissioner::States::Base<Context>
+{
+    AttestationVerification(Context & ctx, Commissionee & commissionee, AttestationInformation & attestationInformation) :
+        Base<Context>(ctx, commissionee, "AttestationVerification"), mAttestationInformation(attestationInformation)
+    {}
+    void Enter()
+    {
+        auto testingRootStore = chip::Credentials::GetTestAttestationTrustStore();
+        auto verifier         = chip::Credentials::GetDefaultDACVerifier(testingRootStore);
+        Callback::Callback<Credentials::OnAttestationInformationVerification> callback(OnVerification, this);
+        verifier->VerifyAttestationInformation(
+            mAttestationInformation.AttestationElements()->Get(), mAttestationInformation.Challenge()->Get(),
+            mAttestationInformation.Signature()->Get(), mAttestationInformation.Pai()->Get(), mAttestationInformation.Dac()->Get(),
+            mAttestationInformation.Nonce()->Get(), &callback);
+    }
+
+    void DispatchSuccess() { this->mCtx.Dispatch(Event::Create<AttestationInformation>(mAttestationInformation)); }
+    void DispatchFailure() { this->mCtx.Dispatch(Event::Create<Failure>()); }
+
+private:
+    static void OnVerification(void * context, Credentials::AttestationVerificationResult result)
+    {
+        auto state = static_cast<AttestationVerification *>(context);
+        if (result == chip::Credentials::AttestationVerificationResult::kSuccess)
+        {
+            state->DispatchSuccess();
+        }
+        else
+        {
+            ChipLogError(Controller,
+                         "Failed in verifying 'Attestation Information' command received from the device: err %hu. Look at "
+                         "AttestationVerificationResult enum to understand the errors",
+                         static_cast<uint16_t>(result));
+            // Go look at AttestationVerificationResult enum in src/credentials/DeviceAttestationVerifier.h to understand the
+            // errors.
+            state->DispatchFailure();
+        }
+    }
+
+    AttestationInformation mAttestationInformation;
+};
+
+struct AttestationVerified : chip::Commissioner::States::Base<Context>
+{
+    AttestationVerified(Context & ctx, Commissionee & commissionee, AttestationInformation & attestationInformation) :
+        Base<Context>(ctx, commissionee, "AttestationVerified"), mAttestationInformation(attestationInformation)
+    {}
+    void Enter()
+    {
+        uint8_t csrNonce[kOpCSRNonceLength];
+        chip::Crypto::DRBG_get_bytes(csrNonce, sizeof(csrNonce));
```

Comment:
This was changed recently (late dec) to use a separate class that creates the nonce and provides the certs.

Reply:
I think you are referring to `CommissioningDelegate.h`, or similar.

This gets to an important feature of the design proposed in this PR.

It is not knowable to the sdk in a general sense where the nonces should be generated.  It is possible that these would be generated in remote infrastructure, which would be required when also remotely verifying attestation signatures.

I know that delegates can be made to provide all of this in an async manner.  However, I believe that these are unwieldy for this purpose.  There are too many interfaces.  I just do not believe that dependency injection is the right solution to this problem and am proposing something different.

Here we're looking at `ExampleCommissioningStateMachine`.  In a product implementation, you can compose your own state machine to do what's needed.  No delegates are required at all.

Reply:
OK, that's fine.  I think you are specifically requesting that the CommissioningParameters interface be preserved in this specific chip-tool implementation.  That's workable.

Reply:
Done.